zkevm-circuits/src/circuit_tools/memory.rs

```diff
@@ -6,230 +6,261 @@ use halo2_proofs::{
     plonk::{Advice, Column, ConstraintSystem, Error, Expression},
     poly::Rotation,
 };
-use std::ops::{Index, IndexMut};
+use std::{
+    collections::HashMap,
+    marker::PhantomData,
+    ops::{Index, IndexMut},
+};
 
 use super::{
-    cached_region::CachedRegion, cell_manager::CellType, constraint_builder::ConstraintBuilder,
+    cached_region::CachedRegion,
+    cell_manager::{CellManager, CellType},
+    constraint_builder::ConstraintBuilder,
 };
 
 #[derive(Clone, Debug, Default)]
-pub(crate) struct Memory<F, C> {
-    columns: Vec<Column<Advice>>,
-    banks: Vec<MemoryBank<F, C>>,
+pub(crate) struct Memory<F: Field, C: CellType, MB: MemoryBank<F, C>> {
+    banks: HashMap<C, MB>,
+    _phantom: PhantomData<F>,
+    tag_counter: usize,
+}
+
+impl<F: Field, C: CellType, MB: MemoryBank<F, C>> Index<C> for Memory<F, C, MB> {
+    type Output = MB;
+
+    fn index(&self, tag: C) -> &Self::Output {
+        if let Some(bank) = self.banks.get(&tag) {
+            bank
+        } else {
+            unreachable!()
+        }
+    }
 }
 
-impl<F: Field, C: CellType> Memory<F, C> {
-    pub(crate) fn new(columns: Vec<Column<Advice>>) -> Self {
+impl<F: Field, C: CellType, MB: MemoryBank<F, C>> IndexMut<C> for Memory<F, C, MB> {
+    fn index_mut(&mut self, tag: C) -> &mut Self::Output {
+        if let Some(bank) = self.banks.get_mut(&tag) {
+            bank
+        } else {
+            unreachable!()
+        }
+    }
+}
+
+impl<F: Field, C: CellType, MB: MemoryBank<F, C>> Memory<F, C, MB> {
+    pub(crate) fn new() -> Self {
         Self {
-            columns,
-            banks: Vec::new(),
+            banks: HashMap::new(),
+            _phantom: PhantomData,
+            tag_counter: 0,
         }
     }
 
-    pub(crate) fn allocate(
+    pub(crate) fn add_rw(
```

Comment:
Yeah you're right, I was trying to find something shorter so that most of the time things still fit on a single line, but not a good enough reason to not use a better name.